prep for making RC4 obfuscator the default

trevj opened the pull request:
This makes the RC4 obfuscator the default. It's mostly a simple bump to our `MESSAGE_VERSION` mechanism which has helped us roll out new features forever. However, I also had to incorporate this mechanism into cloud servers: I added a new command, `version`, which returns the server's `MESSAGE_VERSION`. If it's 6 then the client will offer RC4; otherwise, plain Caesar.

That said, this isn't quite ready yet. I just benchmarked this and am seeing a measurable drop in performance:

| obfuscator | Kb/sec |
| --- | --- |
| none | 1847 |
| caesar | 1798 |
| rc4 | 1482 |

In the real-world any kind of latency between client and server will make this irrelevant but since it really doesn't tally with what I saw in my [previous benchmark](https://github.com/uProxy/uproxy-lib/pull/397) which showed near-identical throughput, I want to investigate a bit.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2508)
<!-- Reviewable:end -->

trevj commented:
I dived into the results a bit more. The previous numbers were from Firefox. Here are a bunch of runs from both Chrome and Firefox:

Firefox:
```
** raw numbers: {'none': 1874, 'caesar': 1854, 'rc4': 1641}
** raw numbers: {'none': 1954, 'caesar': 1855, 'rc4': 1635}
** raw numbers: {'none': 1964, 'caesar': 1834, 'rc4': 1627}
```

Chrome:
```
** raw numbers: {'none': 1909, 'caesar': 1805, 'rc4': 1602}
** raw numbers: {'none': 1882, 'caesar': 1822, 'rc4': 1641}
** raw numbers: {'none': 1883, 'caesar': 1807, 'rc4': 1630}
```

It's basically a 10% drop vs. Caesar - not great but given how much less detectable it is, I'd take it.

That said, I quickly ran the numbers and I'm seeing a large drop in performance in Firefox when I introduce 150ms of latency between the two pairs (much more representative of real-world conditions):
```
Chrome: ** raw numbers: {'caesar': 524, 'rc4': 521}
Firefox: ** raw numbers: {'caesar': 525, 'rc4': 256}

```

So, I still need to investigate! My guess is the random number buffering stuff is slower on FF - I will try with newer versions of FF (IIRC, 48 introduces access to random numbers in web workers).

mollyling commented:
Avg for Firefox: rc4: 1634.3, caesar: 1847.6, rc4 is 88.46% the speed of
caesar.
Avg for Chrome: rc4: 1624.3, caesar: 1811.3, rc4 is 89.68% the speed of
caesar.

That's pretty good!  Considering that you can't just create 255 table
entries to byte-match rc4 the way you can caesar...

Is MESSAGE_VERSION the only way to determine the protocol between two
peers?  What about a list of keywords (in prio order)?



Cheers,
$ ls

Lally Singh |  Jigsaw <https://jigsaw.google.com> |  Engineering |
lally@google.com


On Thu, Jun 23, 2016 at 12:44 PM, trevj <notifications@github.com> wrote:

> This makes the RC4 obfuscator the default. It's mostly a simple bump to
> our MESSAGE_VERSION mechanism which has helped us roll out new features
> forever. However, I also had to incorporate this mechanism into cloud
> servers: I added a new command, version, which returns the server's
> MESSAGE_VERSION. If it's 6 then the client will offer RC4; otherwise,
> plain Caesar.
>
> That said, this isn't quite ready yet. I just benchmarked this and am
> seeing a measurable drop in performance:
>
> | obfuscator | Kb/sec |
> | -- | -- |
> | none | 1847 |
> | caesar | 1798 |
> | rc4 | 1482 |
>
> In the real-world any kind of latency between client and server will make
> this irrelevant but since it really doesn't tally with what I saw in my previous
> benchmark <https://github.com/uProxy/uproxy-lib/pull/397> which showed
> near-identical throughput, I want to investigate a bit.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/uProxy/uproxy/pull/2508
> Commit Summary
>
>    - update MESSAGE_VERSION changelog
>    - add a MESSAGE_VERSION for RC4 obfuscation
>    - mention that MESSAGE_VERSION=5 has been obsoleted
>    - move pure constants out of globals.ts so they can be easily included
>    by other apps, e.g. zork
>    - add a version command to zork so that MESSAGE_VERSION can be
>    determined
>    - query server MESSAGE_VERSION on login so client can determine which
>    features the server supports
>    - use a cryptographically-secure random config for RC4 connections
>
> File Changes
>
>    - *A* src/generic_core/constants.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-0> (23)
>    - *M* src/generic_core/globals.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-1> (32)
>    - *M* src/generic_core/remote-connection.spec.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-2> (20)
>    - *M* src/generic_core/remote-connection.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-3> (15)
>    - *M* src/generic_core/remote-instance.spec.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-4> (17)
>    - *M* src/generic_core/remote-user.spec.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-5> (11)
>    - *M* src/generic_core/social.spec.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-6> (6)
>    - *M* src/generic_core/uproxy_core.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-7> (5)
>    - *M* src/lib/cloud/social/provider.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-8> (64)
>    - *M* src/lib/transformers/rc4.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-9> (7)
>    - *M* src/lib/zork/freedom-module.ts
>    <https://github.com/uProxy/uproxy/pull/2508/files#diff-10> (5)
>
> Patch Links:
>
>    - https://github.com/uProxy/uproxy/pull/2508.patch
>    - https://github.com/uProxy/uproxy/pull/2508.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/uProxy/uproxy/pull/2508>, or mute the thread
> <https://github.com/notifications/unsubscribe/AEuYJtciTVfnUvPudup8ogjlpyic2pJ6ks5qOrf6gaJpZM4I9AoH>
> .
>

trevj commented:
@mollyling Thanks! Yes I think that's acceptable given how much harder it should be to detect...as for `MESSAGE_VERSION` we absolutely 100% will have to think about offering multiple obfuscators in one session but for right now `MESSAGE_VERSION` should work fine.

trevj commented:
TL;DR: RC4 obfuscator is slow in Firefox versions <48, which (finally) allows web workers in add-ons access to `window.crypto`

Quick figures:
firefox stable (47): {'caesar': 526, 'rc4': 344}
firefox beta (48): {'caesar': 529, 'rc4': 523}
firefox canary (49): {'caesar': 523, 'rc4': 520}

Given this, it may be prudent to wait until Firefox 48 becomes stable before deploying - mostly I'm thinking of cloud servers. That's slated for the first week of August:
https://wiki.mozilla.org/RapidRelease/Calendar

trevj commented:
@soycode thought you might find this interesting...I've no guesses as to why the random buffering stuff would be so much slower **even** with high latency between the two peers

soycode commented:
Hrm, I'm not sure - the randomness fill buffers in a probably wastefully naive way, so that could be part of it. It's a little tighter in the pgp module since it's easier to know when to refill the buffer.

trevj commented:
Yeah...and I don't want to spend much time optimising it now. If it was like one hour of work sure but I don't really have any great ideas on how to improve it.

soycode commented:
Assuming it is the buffer, the fix would basically be:
- Expose a `refillBuffer()` method
- Have the client call it when needed instead of doing it inside the code itself

A bit ugly as it'll need a bunch of `if (firefox) {...}` sort of code in the client. Exposing the same simple API lets the client be unchanged, but yeah likely results in this inefficiency.

trevj commented:
OK, let's hold off on making RC4 the default for now. I added TODO: to `constants.ts`. I'd like to get all the plumbing out and under testing.

trevj commented:
@jpevarnek any thoughts on the `constants.ts`/`globals.ts` change?

jab commented on src/generic_core/remote-connection.ts:
```diff
@@ -296,12 +297,20 @@ var generateProxyingSessionId_ = (): string => {
           pc = bridge.preObfuscation('sockstortc', config, this.portControl_);
           break;
         case 3:
-          log.debug('using bridge with basicObfuscation');
```
Would it be clearer and less error-prone to use an enum or constant here, instead of these numeric literals?

trevj replied:
Yes, it would. I introduced an enum, `MESSAGE_VERSIONS`.

jab commented on src/generic_core/remote-connection.ts:
```diff
@@ -296,12 +297,20 @@ var generateProxyingSessionId_ = (): string => {
+          log.debug('using bridge with caesar obfuscation');
           pc = bridge.basicObfuscation('sockstortc', config, this.portControl_);
           break;
+        case 4:
```
Some people advise adding a comment like `// fall through to case BAZ` when falling through is intended just to make it super clear it's not an oversight, but totes up to you. (I'm not one of those zealots who argues that switch/case statements are too error-prone to even ever use, but I can see where they're coming from.)

trevj replied:
Excellent idea. I've added a comment.

jab commented:
Just took a quick look, so sorry again I didn't get to it sooner. 😢 
Only familiar enough to give a superficial 👍  (modulo my superficial comments). Probably worth getting @bemasc's 👍  before merging.

jpevarnek commented:
:thumbsup: with small questions

---

Reviewed 5 of 11 files at r1, 1 of 2 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/generic_core/remote-connection.ts, line 304 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2508#-KN3_-Cqwn1I3lKwxCzX:-KN3_-Cqwn1I3lKwxCzY:-1275552403) ([raw file](https://github.com/uproxy/uproxy/blob/cf08f19d598e49d66a8911e077d3e5dd6b8bee40/src/generic_core/remote-connection.ts#L304)):*
> ```XML
>           pc = bridge.basicObfuscation('sockstortc', config, this.portControl_);
>           break;
>         case 4:
> ```

I'm a bit out of the loop on backend changes, what were versions 4 and 5 and what is different with them?

---

*[src/lib/cloud/social/provider.ts, line 199 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2508#-KN3_S8xhFyzdNRzttAY:-KN3_S8ywVO59LyG0s9c:-399787761) ([raw file](https://github.com/uproxy/uproxy/blob/cf08f19d598e49d66a8911e077d3e5dd6b8bee40/src/lib/cloud/social/provider.ts#L199)):*
> ```XML
>       from: clientState,
>       // INSTANCE
>       message: JSON.stringify(makeVersionedPeerMessage(3000, makeInstanceMessage(
> ```

Do we have a constant for the `3000`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2508#-:-KN3_ogYpRgbMJ2_JZju:-690062835)*
<!-- Sent from Reviewable.io -->

trevj commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[src/generic_core/remote-connection.ts, line 303 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2508#-KKy8YwOQmaPvkXYGHRf-r2-303:-KN8yx4A-q_nCTBtsAqy:271672073) ([raw file](https://github.com/uproxy/uproxy/blob/1462744e2b7c3fded8e842db5747e1308779d362/src/generic_core/remote-connection.ts#L303)):*
<details><summary><i>Previously, jab wrote…</i></summary>
> Some people advise adding a comment like `// fall through to case BAZ` when falling through is intended just to make it super clear it's not an oversight, but totes up to you. (I'm not one of those zealots who argues that switch/case statements are too error-prone to even ever use, but I can see where they're coming from.)
</details>
Excellent idea. I've added a comment.

---

*[src/generic_core/remote-connection.ts, line 304 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2508#-KN3_-Cqwn1I3lKwxCzX:-KN8y7mB9V_K-EHVIGqR:-1173318527) ([raw file](https://github.com/uproxy/uproxy/blob/cf08f19d598e49d66a8911e077d3e5dd6b8bee40/src/generic_core/remote-connection.ts#L304)):*
<details><summary><i>Previously, jpevarnek (Jonathan Pevarnek) wrote…</i></summary>
> I'm a bit out of the loop on backend changes, what were versions 4 and 5 and what is different with them?
</details>
4 and 5 were Holographic ICE with caesar obfuscation; 6 adds RC4 obfuscation. I've added a note to consult `globals.ts` for information on each version.

---

*[src/lib/cloud/social/provider.ts, line 199 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2508#-KN3_S8xhFyzdNRzttAY:-KN8zcubHW5Ka-XaVeWH:-1471425417) ([raw file](https://github.com/uproxy/uproxy/blob/cf08f19d598e49d66a8911e077d3e5dd6b8bee40/src/lib/cloud/social/provider.ts#L199)):*
<details><summary><i>Previously, jpevarnek (Jonathan Pevarnek) wrote…</i></summary>
> Do we have a constant for the `3000`?
</details>
You know, we do - you've reminded me to use typings (trivial now, since we moved this code to the social provider):
https://github.com/uProxy/uproxy/issues/2600

Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2508)*
<!-- Sent from Reviewable.io -->
